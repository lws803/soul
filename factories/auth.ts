import { Factory } from 'fishery';

import { RefreshToken } from 'src/auth/entities/refresh-token.entity';
import { JWTRefreshPayload } from 'src/auth/entities/jwt-refresh-payload.entity';
import { JWTPayload } from 'src/auth/entities/jwt-payload.entity';
import { TokenType } from 'src/auth/enums/token-type.enum';

import * as factories from './index';

export const refreshToken = Factory.define<RefreshToken>(() => ({
  id: 1,
  user: factories.user.build(),
  isRevoked: false,
  createdAt: new Date('1995-12-17T03:24:00'),
  updatedAt: new Date('1995-12-18T03:24:00'),
  expires: new Date('1995-12-19T03:24:00'),
  platformUser: factories.platformUser.build(),
}));

export const jwtRefreshPayload = Factory.define<JWTRefreshPayload>(() => ({
  tokenId: refreshToken.build().id,
  userId: factories.user.build().id,
  tokenType: TokenType.Refresh,
}));

export const jwtPayload = Factory.define<JWTPayload>(() => {
  const oneUser = factories.user.build();
  return {
    userId: oneUser.id,
    username: oneUser.username,
    tokenType: TokenType.Access,
  };
});

export const jwtPayloadWithPlatform = Factory.define<JWTPayload>(() => {
  const oneUser = factories.user.build();
  const onePlatformUser = factories.platformUser.build({ user: oneUser });
  return {
    userId: oneUser.id,
    username: oneUser.username,
    tokenType: TokenType.Access,
    platformId: onePlatformUser.platform.id,
    roles: onePlatformUser.roles,
  };
});

export const jwtRefreshPayloadWithPlatform = Factory.define<JWTRefreshPayload>(
  () => {
    const oneUser = factories.user.build();
    const onePlatformUser = factories.platformUser.build({ user: oneUser });
    return {
      tokenId: refreshToken.build().id,
      userId: oneUser.id,
      tokenType: TokenType.Refresh,
      platformId: onePlatformUser.platform.id,
      roles: onePlatformUser.roles,
    };
  },
);

export const requestUserObject = Factory.define<{
  tokenType: TokenType;
  username: string;
  userId: number;
}>(() => ({
  tokenType: TokenType.Access,
  username: 'TEST_USER',
  userId: 1,
}));
