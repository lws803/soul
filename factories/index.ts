export * from './users';
export * from './auth';
export * from './platform';
export * from './user-connections';
