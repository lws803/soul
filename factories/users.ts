import { Factory } from 'fishery';

import { CreateUserDto, UpdateUserDto } from 'src/users/serializers/api.dto';
import { User } from 'src/users/entities/user.entity';

export const user = Factory.define<User>(({ sequence }) => {
  // Rewinds the sequence as we do not want to have "sequences" bleed into subsequent builds
  user.rewindSequence();
  return {
    id: sequence,
    username: `test-user-${sequence}`,
    userHandle: `test-user-${sequence}#${sequence}`,
    email: `TEST_USER_${sequence}@EMAIL.COM`,
    isActive: true,
    createdAt: new Date('1995-12-17T03:24:00'),
    updatedAt: new Date('1995-12-18T03:24:00'),
    hashedPassword: 'TEST_HASHED_PASSWORD',
    displayName: 'DISPLAY_NAME',
    bio: 'BIO',
  };
});

export const createUserDto = Factory.define<CreateUserDto>(() => ({
  email: 'TEST_USER@EMAIL.COM',
  username: 'test-user',
  password: 'TEST_PASSWORD',
  bio: null,
  displayName: null,
}));

export const updateUserDto = Factory.define<UpdateUserDto>(() => ({
  email: 'UPDATED_EMAIL@EMAIL.COM',
  username: 'updated-user',
  displayName: 'UPDATED_DISPLAY_NAME',
  bio: 'UPDATED_BIO',
}));
