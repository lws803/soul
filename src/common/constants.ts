export const DEFAULT_NUM_ITEMS_PER_PAGE = 10;
