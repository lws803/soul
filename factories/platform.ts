import { Factory } from 'fishery';

import * as factories from 'factories';
import { Platform } from 'src/platforms/entities/platform.entity';
import { PlatformUser } from 'src/platforms/entities/platform-user.entity';
import {
  UpdatePlatformDto,
  CreatePlatformDto,
} from 'src/platforms/dto/api.dto';
import { UserRole } from 'src/roles/role.enum';
import { PlatformCategory } from 'src/platforms/entities/platform-category.entity';

export const onePlatformCategory = Factory.define<PlatformCategory>(() => ({
  id: 1,
  name: 'CATEGORY',
}));

export const onePlatform = Factory.define<Platform>(() => ({
  id: 1,
  name: 'TEST_PLATFORM',
  nameHandle: 'test_platform#1',
  createdAt: new Date('1995-12-17T03:24:00'),
  updatedAt: new Date('1995-12-18T03:24:00'),
  userConnections: [],
  isVerified: true,
  redirectUris: ['TEST_REDIRECT_URI'],
  category: onePlatformCategory.build(),
  activityWebhookUri: 'ACTIVITY_WEBHOOK_URI',
}));

export const platformArray = Factory.define<Platform[]>(() => [
  onePlatform.build(),
  onePlatform.build({ id: 2, name: 'TEST_PLATFORM_2' }),
]);

export const onePlatformUser = Factory.define<PlatformUser>(() => ({
  id: 1,
  user: factories.oneUser.build(),
  platform: factories.onePlatform.build(),
  roles: [UserRole.Admin, UserRole.Member],
  createdAt: new Date('1995-12-17T03:24:00'),
  updatedAt: new Date('1995-12-18T03:24:00'),
}));

export const platformUserArray = Factory.define<PlatformUser[]>(() => [
  onePlatformUser.build(),
  onePlatformUser.build({ id: 2 }),
]);

export const updatePlatformDto = Factory.define<UpdatePlatformDto>(() => ({
  name: 'TEST_PLATFORM_UPDATE',
  category: 'CATEGORY_UPDATE',
}));

export const createPlatformDto = Factory.define<CreatePlatformDto>(() => ({
  name: 'TEST_PLATFORM',
  redirectUris: ['TEST_REDIRECT_URI'],
  category: onePlatformCategory.build().name,
}));

type CreatePlatformRequest = {
  name: string;
  redirect_uris: string[];
  category: string;
};

export const createPlatformRequestDto = Factory.define<CreatePlatformRequest>(
  () => ({
    name: 'TEST_PLATFORM',
    redirect_uris: ['TEST_REDIRECT_URI'],
    category: onePlatformCategory.build().name,
  }),
);
