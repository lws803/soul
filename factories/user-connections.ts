import { Factory } from 'fishery';

import { FindOneUserConnectionResponseEntity } from 'src/user-connections/serializers/api-responses.entity';

import { UserConnection } from 'src/user-connections/entities/user-connection.entity';

import { platformEntity } from './platform';
import { userEntity } from './users';

export const userConnectionEntity = Factory.define<UserConnection>(() => ({
  id: 1,
  fromUser: userEntity.build(),
  toUser: userEntity.build({ id: 2, email: 'TEST_USER_2@EMAIL.COM' }),
  platforms: [],
  createdAt: new Date('1995-12-17T03:24:00'),
  updatedAt: new Date('1995-12-18T03:24:00'),
  mutualConnection: null,
}));

export const oneUserConnectionResponseEntity =
  Factory.define<FindOneUserConnectionResponseEntity>(() => ({
    id: 1,
    fromUser: userEntity.build(),
    toUser: userEntity.build({ id: 2, email: 'TEST_USER_2@EMAIL.COM' }),
    platforms: [],
    createdAt: new Date('1995-12-17T03:24:00'),
    updatedAt: new Date('1995-12-18T03:24:00'),
    mutualConnection: null,
    isMutual: false,
  }));

export const userConnectionEntityArray = Factory.define<UserConnection[]>(
  () => [
    userConnectionEntity.build({ id: 1 }),
    userConnectionEntity.build({
      id: 2,
      fromUser: userEntity.build({ id: 3 }),
      toUser: userEntity.build({ id: 4 }),
    }),
  ],
);

type CreateUserConnectionRequest = {
  to_user_id: number;
  platform_id?: number;
};

export const createUserConnectionRequest =
  Factory.define<CreateUserConnectionRequest>(() => ({
    to_user_id: userEntity.build({ id: 2 }).id,
  }));

type PostPlatformToUserConnectionRequest = {
  platform_id: number;
};

export const postPlatformToUserConnectionRequest =
  Factory.define<PostPlatformToUserConnectionRequest>(() => ({
    platform_id: platformEntity.build().id,
  }));
