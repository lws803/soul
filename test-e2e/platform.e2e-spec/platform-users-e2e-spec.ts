import { HttpStatus, INestApplication } from '@nestjs/common';
import * as request from 'supertest';
import { Repository, Connection } from 'typeorm';
import * as sha256 from 'crypto-js/sha256';
import base64url from 'base64url';

import { UserRole } from 'src/roles/role.enum';
import { PlatformUser } from 'src/platforms/entities/platform-user.entity';
import { Platform } from 'src/platforms/entities/platform.entity';
import { PlatformCategory } from 'src/platforms/entities/platform-category.entity';

import * as factories from '../../factories';
import createAppFixture from '../fixtures/create-app-fixture';
import {
  createUsersAndLoginFixture,
  UserAccount,
} from '../fixtures/create-users-and-login-fixture';

describe('PlatformsController - PlatformUsers (e2e)', () => {
  let app: INestApplication;
  let platformUserRepository: Repository<PlatformUser>;
  let platformRepository: Repository<Platform>;
  let platformCategoryRepository: Repository<PlatformCategory>;

  let userAccount: UserAccount;
  let secondUserAccount: UserAccount;
  let thirdUserAccount: UserAccount;

  const codeVerifier = 'CODE_VERIFIER';
  const codeChallenge = base64url(sha256(codeVerifier).toString(), 'hex');

  beforeAll(async () => {
    app = await createAppFixture({});
    await app.init();
    app.useLogger(false);

    const connection = app.get(Connection);
    await connection.synchronize(true);

    platformUserRepository = connection.getRepository(PlatformUser);
    platformRepository = connection.getRepository(Platform);
    platformCategoryRepository = connection.getRepository(PlatformCategory);

    const [firstUser, secondUser, thirdUser] = await createUsersAndLoginFixture(
      app,
    );
    userAccount = firstUser;
    secondUserAccount = secondUser;
    thirdUserAccount = thirdUser;

    await platformCategoryRepository.save(
      factories.platformCategoryEntity.build(),
    );
  });

  afterAll(async () => {
    await app.close();
  });

  describe('/platforms/:platform_id/users (GET)', () => {
    beforeAll(async () => {
      const platform = await platformRepository.save(
        factories.platformEntity.build({
          redirectUris: ['https://www.example.com'],
          clientSecret: 'CLIENT_SECRET',
        }),
      );
      await platformUserRepository.save([
        factories.platformUserEntity.build({
          user: userAccount.user,
          platform,
        }),
        factories.platformUserEntity.build({
          id: 2,
          user: secondUserAccount.user,
          platform,
        }),
        factories.platformUserEntity.build({
          id: 3,
          user: thirdUserAccount.user,
          platform,
        }),
      ]);
    });

    afterAll(async () => {
      await platformUserRepository.delete({});
      await platformRepository.delete({});
    });

    it('fetches all users within a platform', async () => {
      const response = await request(app.getHttpServer())
        .post('/auth/oauth/client-credentials')
        .send({
          client_secret: 'CLIENT_SECRET',
          client_id: 1,
        });

      await request(app.getHttpServer())
        .get('/platforms/1/users')
        .set('Authorization', `Bearer ${response.body.access_token}`)
        .expect(HttpStatus.OK)
        .expect((res) =>
          expect(res.body).toEqual({
            total_count: 3,
            platform_users: [
              {
                id: expect.any(Number),
                roles: [UserRole.Admin, UserRole.Member],
                profile_url: 'PROFILE_URL',
                user: {
                  id: expect.any(Number),
                  user_handle: 'test-user#1',
                  username: 'test-user',
                  bio: null,
                  display_name: null,
                  email: 'TEST_USER@EMAIL.COM',
                  is_active: true,
                  created_at: expect.any(String),
                  updated_at: expect.any(String),
                },
              },
              {
                id: expect.any(Number),
                roles: [UserRole.Admin, UserRole.Member],
                profile_url: 'PROFILE_URL',
                user: {
                  id: expect.any(Number),
                  user_handle: 'test-user-2#2',
                  username: 'test-user-2',
                  bio: null,
                  display_name: null,
                  email: 'TEST_USER_2@EMAIL.COM',
                  is_active: true,
                  created_at: expect.any(String),
                  updated_at: expect.any(String),
                },
              },
              {
                id: expect.any(Number),
                roles: [UserRole.Admin, UserRole.Member],
                profile_url: 'PROFILE_URL',
                user: {
                  id: expect.any(Number),
                  user_handle: 'test-user-3#3',
                  username: 'test-user-3',
                  bio: null,
                  display_name: null,
                  email: 'TEST_USER_3@EMAIL.COM',
                  is_active: true,
                  created_at: expect.any(String),
                  updated_at: expect.any(String),
                },
              },
            ],
          }),
        );
    });

    it('fetches all users within a platform with filters', async () => {
      const response = await request(app.getHttpServer())
        .post('/auth/oauth/client-credentials')
        .send({
          client_secret: 'CLIENT_SECRET',
          client_id: 1,
        });

      await request(app.getHttpServer())
        .get('/platforms/1/users?uid=1&uid=2')
        .set('Authorization', `Bearer ${response.body.access_token}`)
        .expect(HttpStatus.OK)
        .expect((res) =>
          expect(res.body).toEqual({
            total_count: 2,
            platform_users: [
              {
                id: expect.any(Number),
                roles: [UserRole.Admin, UserRole.Member],
                profile_url: 'PROFILE_URL',
                user: {
                  id: expect.any(Number),
                  user_handle: 'test-user#1',
                  username: 'test-user',
                  bio: null,
                  display_name: null,
                  email: 'TEST_USER@EMAIL.COM',
                  is_active: true,
                  created_at: expect.any(String),
                  updated_at: expect.any(String),
                },
              },
              {
                id: expect.any(Number),
                roles: [UserRole.Admin, UserRole.Member],
                profile_url: 'PROFILE_URL',
                user: {
                  id: expect.any(Number),
                  user_handle: 'test-user-2#2',
                  username: 'test-user-2',
                  bio: null,
                  display_name: null,
                  email: 'TEST_USER_2@EMAIL.COM',
                  is_active: true,
                  created_at: expect.any(String),
                  updated_at: expect.any(String),
                },
              },
            ],
          }),
        );
    });
  });

  describe('/platforms/:platform_id/users/:user_id (GET)', () => {
    beforeAll(async () => {
      const platform = await platformRepository.save(
        factories.platformEntity.build({
          redirectUris: ['https://www.example.com'],
          clientSecret: 'CLIENT_SECRET',
        }),
      );
      await platformUserRepository.save([
        factories.platformUserEntity.build({
          user: userAccount.user,
          platform,
        }),
        factories.platformUserEntity.build({
          id: 2,
          user: secondUserAccount.user,
          platform,
        }),
        factories.platformUserEntity.build({
          id: 3,
          user: thirdUserAccount.user,
          platform,
        }),
      ]);
    });

    afterAll(async () => {
      await platformUserRepository.delete({});
      await platformRepository.delete({});
    });

    it('fetches single user within a platform', async () => {
      const response = await request(app.getHttpServer())
        .post('/auth/oauth/client-credentials')
        .send({
          client_secret: 'CLIENT_SECRET',
          client_id: 1,
        });

      await request(app.getHttpServer())
        .get('/platforms/1/users/1')
        .set('Authorization', `Bearer ${response.body.access_token}`)
        .expect(HttpStatus.OK)
        .expect((res) =>
          expect(res.body).toEqual({
            id: expect.any(Number),
            roles: [UserRole.Admin, UserRole.Member],
            profile_url: 'PROFILE_URL',
            user: {
              id: expect.any(Number),
              user_handle: 'test-user#1',
              username: 'test-user',
              bio: null,
              display_name: null,
              email: 'TEST_USER@EMAIL.COM',
              is_active: true,
              created_at: expect.any(String),
              updated_at: expect.any(String),
            },
          }),
        );
    });

    it('throws not found when user is not found', async () => {
      const response = await request(app.getHttpServer())
        .post('/auth/oauth/client-credentials')
        .send({
          client_secret: 'CLIENT_SECRET',
          client_id: 1,
        });

      await request(app.getHttpServer())
        .get('/platforms/1/users/999')
        .set('Authorization', `Bearer ${response.body.access_token}`)
        .expect(HttpStatus.NOT_FOUND);
    });
  });

  describe('/platforms/:platform_id/users/:user_id (PATCH)', () => {
    beforeEach(async () => {
      const platform = await platformRepository.save(
        factories.platformEntity.build({
          redirectUris: ['https://www.example.com'],
          clientSecret: 'CLIENT_SECRET',
        }),
      );
      await platformUserRepository.save([
        factories.platformUserEntity.build({
          user: userAccount.user,
          platform,
        }),
        factories.platformUserEntity.build({
          id: 2,
          user: secondUserAccount.user,
          platform,
          roles: [UserRole.Member],
        }),
      ]);
    });

    afterEach(async () => {
      await platformUserRepository.delete({});
      await platformRepository.delete({});
    });

    it('sets user role', async () => {
      const response = await request(app.getHttpServer())
        .post('/auth/oauth/client-credentials')
        .send({
          client_secret: 'CLIENT_SECRET',
          client_id: 1,
        });

      await request(app.getHttpServer())
        .patch('/platforms/1/users/2')
        .set('Authorization', `Bearer ${response.body.access_token}`)
        .send({
          ...factories.updatePlatformUserRequest.build(),
          profile_url: undefined,
        })
        .expect(HttpStatus.OK)
        .expect((res) =>
          expect(res.body).toEqual({
            id: 2,
            profile_url: 'PROFILE_URL',
            roles: [UserRole.Admin, UserRole.Member],
            user: {
              id: 2,
              user_handle: 'test-user-2#2',
              username: 'test-user-2',
              bio: null,
              display_name: null,
              email: 'TEST_USER_2@EMAIL.COM',
              is_active: true,
              created_at: expect.any(String),
              updated_at: expect.any(String),
            },
          }),
        );
    });

    it('sets user role and profile url', async () => {
      const response = await request(app.getHttpServer())
        .post('/auth/oauth/client-credentials')
        .send({
          client_secret: 'CLIENT_SECRET',
          client_id: 1,
        });

      await request(app.getHttpServer())
        .patch('/platforms/1/users/2')
        .set('Authorization', `Bearer ${response.body.access_token}`)
        .send(factories.updatePlatformUserRequest.build())
        .expect(HttpStatus.OK)
        .expect((res) =>
          expect(res.body).toEqual({
            id: 2,
            profile_url: 'PROFILE_URL_UPDATE',
            roles: [UserRole.Admin, UserRole.Member],
            user: {
              id: 2,
              user_handle: 'test-user-2#2',
              username: 'test-user-2',
              bio: null,
              display_name: null,
              email: 'TEST_USER_2@EMAIL.COM',
              is_active: true,
              created_at: expect.any(String),
              updated_at: expect.any(String),
            },
          }),
        );
    });

    it('sets user profile url', async () => {
      const response = await request(app.getHttpServer())
        .post('/auth/oauth/client-credentials')
        .send({
          client_secret: 'CLIENT_SECRET',
          client_id: 1,
        });

      await request(app.getHttpServer())
        .patch('/platforms/1/users/2')
        .set('Authorization', `Bearer ${response.body.access_token}`)
        .send({
          ...factories.updatePlatformUserRequest.build(),
          roles: undefined,
        })
        .expect(HttpStatus.OK)
        .expect((res) =>
          expect(res.body).toEqual({
            id: 2,
            profile_url: 'PROFILE_URL_UPDATE',
            roles: [UserRole.Member],
            user: {
              id: 2,
              user_handle: 'test-user-2#2',
              username: 'test-user-2',
              bio: null,
              display_name: null,
              email: 'TEST_USER_2@EMAIL.COM',
              is_active: true,
              created_at: expect.any(String),
              updated_at: expect.any(String),
            },
          }),
        );
    });

    it('removes user profile url', async () => {
      const response = await request(app.getHttpServer())
        .post('/auth/oauth/client-credentials')
        .send({
          client_secret: 'CLIENT_SECRET',
          client_id: 1,
        });

      await request(app.getHttpServer())
        .patch('/platforms/1/users/2')
        .set('Authorization', `Bearer ${response.body.access_token}`)
        .send({
          ...factories.updatePlatformUserRequest.build({ profile_url: null }),
          roles: undefined,
        })
        .expect(HttpStatus.OK)
        .expect((res) =>
          expect(res.body).toEqual({
            id: 2,
            profile_url: null,
            roles: [UserRole.Member],
            user: {
              id: 2,
              user_handle: 'test-user-2#2',
              username: 'test-user-2',
              bio: null,
              display_name: null,
              email: 'TEST_USER_2@EMAIL.COM',
              is_active: true,
              created_at: expect.any(String),
              updated_at: expect.any(String),
            },
          }),
        );
    });

    it('bans a user', async () => {
      const response = await request(app.getHttpServer())
        .post('/auth/oauth/client-credentials')
        .send({
          client_secret: 'CLIENT_SECRET',
          client_id: 1,
        });

      await request(app.getHttpServer())
        .patch('/platforms/1/users/2')
        .set('Authorization', `Bearer ${response.body.access_token}`)
        .send({
          ...factories.updatePlatformUserRequest.build({ roles: ['banned'] }),
          profile_url: undefined,
        })
        .expect(HttpStatus.OK)
        .expect((res) =>
          expect(res.body).toEqual({
            id: 2,
            profile_url: 'PROFILE_URL',
            roles: [UserRole.Banned],
            user: {
              id: 2,
              user_handle: 'test-user-2#2',
              username: 'test-user-2',
              bio: null,
              display_name: null,
              email: 'TEST_USER_2@EMAIL.COM',
              is_active: true,
              created_at: expect.any(String),
              updated_at: expect.any(String),
            },
          }),
        );
    });

    it('throws an error when trying to set only remaining admin to member', async () => {
      const response = await request(app.getHttpServer())
        .post('/auth/oauth/client-credentials')
        .send({
          client_secret: 'CLIENT_SECRET',
          client_id: 1,
        });

      await request(app.getHttpServer())
        .patch('/platforms/1/users/1')
        .set('Authorization', `Bearer ${response.body.access_token}`)
        .send({
          ...factories.updatePlatformUserRequest.build({ roles: ['member'] }),
          profile_url: undefined,
        })
        .expect(HttpStatus.FORBIDDEN)
        .expect((res) =>
          expect(res.body).toEqual({
            error: 'NO_ADMINS_REMAINING',
            message:
              'It seems like you might be the last admin of this platform. ' +
              'You need to appoint another admin before performing this action.',
          }),
        );
    });
  });

  describe('/platforms/:platform_id/users/:user_id (DELETE)', () => {
    beforeEach(async () => {
      const platform = await platformRepository.save(
        factories.platformEntity.build({
          redirectUris: ['https://www.example.com'],
          clientSecret: 'CLIENT_SECRET',
        }),
      );
      await platformUserRepository.save([
        factories.platformUserEntity.build({
          user: userAccount.user,
          platform,
        }),
        factories.platformUserEntity.build({
          id: 2,
          user: secondUserAccount.user,
          platform,
          roles: [UserRole.Member],
        }),
      ]);
    });

    afterEach(async () => {
      await platformUserRepository.delete({});
      await platformRepository.delete({});
    });

    it('deletes a platform user', async () => {
      const response = await request(app.getHttpServer())
        .post('/auth/oauth/client-credentials')
        .send({
          client_secret: 'CLIENT_SECRET',
          client_id: 1,
        });

      await request(app.getHttpServer())
        .delete('/platforms/1/users/2')
        .set('Authorization', `Bearer ${response.body.access_token}`)
        .expect(HttpStatus.OK)
        .expect((res) => expect(res.body).toEqual({}));
    });
  });

  describe('/platforms/:platform_id/quit (DELETE)', () => {
    let platform: Platform;

    beforeEach(async () => {
      platform = await platformRepository.save(
        factories.platformEntity.build({
          redirectUris: ['https://www.example.com'],
        }),
      );

      await platformUserRepository.save([
        factories.platformUserEntity.build({
          user: userAccount.user,
          platform,
        }),
        factories.platformUserEntity.build({
          id: 2,
          user: secondUserAccount.user,
          platform,
          roles: [UserRole.Member],
        }),
        factories.platformUserEntity.build({
          id: 3,
          user: thirdUserAccount.user,
          platform,
          roles: [UserRole.Banned],
        }),
      ]);
    });

    afterEach(async () => {
      await platformUserRepository.delete({});
      await platformRepository.delete({});
    });

    it('quits existing platform (ADMIN)', async () => {
      await platformUserRepository.save(
        factories.platformUserEntity.build({
          id: 2,
          user: secondUserAccount.user,
          platform,
          roles: [UserRole.Admin, UserRole.Member],
        }),
      );

      const params = new URLSearchParams({
        redirect_uri: 'https://www.example.com',
        state: 'TEST_STATE',
        code_challenge: codeChallenge,
        client_id: String(1),
      });
      const codeResp = await request(app.getHttpServer())
        .post(`/auth/code?${params.toString()}`)
        .send({ email: 'TEST_USER@EMAIL.COM', password: '1oNc0iY3oml5d&%9' });
      const response = await request(app.getHttpServer())
        .post('/auth/oauth/authorization-code')
        .send({
          code: codeResp.body.code,
          redirect_uri: 'https://www.example.com',
          code_verifier: codeVerifier,
        });

      await request(app.getHttpServer())
        .delete('/platforms/1/quit')
        .set('Authorization', `Bearer ${response.body.access_token}`)
        .expect(HttpStatus.OK)
        .expect((res) => expect(res.body).toEqual({}));
    });

    it('quits existing platform (MEMBER)', async () => {
      const params = new URLSearchParams({
        redirect_uri: 'https://www.example.com',
        state: 'TEST_STATE',
        code_challenge: codeChallenge,
        client_id: String(1),
      });
      const codeResp = await request(app.getHttpServer())
        .post(`/auth/code?${params.toString()}`)
        .send({ email: 'TEST_USER_2@EMAIL.COM', password: '1oNc0iY3oml5d&%9' });
      const response = await request(app.getHttpServer())
        .post('/auth/oauth/authorization-code')
        .send({
          code: codeResp.body.code,
          redirect_uri: 'https://www.example.com',
          code_verifier: codeVerifier,
        });

      await request(app.getHttpServer())
        .delete('/platforms/1/quit')
        .set('Authorization', `Bearer ${response.body.access_token}`)
        .expect(HttpStatus.OK)
        .expect((res) => expect(res.body).toEqual({}));
    });

    // Banned users cannot quit a platform to prevent them from joining again
    it('quits existing platform (BANNED)', async () => {
      const response = await request(app.getHttpServer())
        .post('/auth/login')
        .send({ email: 'TEST_USER_3@EMAIL.COM', password: '1oNc0iY3oml5d&%9' });

      await request(app.getHttpServer())
        .delete('/platforms/1/quit')
        .set('Authorization', `Bearer ${response.body.access_token}`)
        .expect(HttpStatus.FORBIDDEN)
        .expect((res) =>
          expect(res.body).toEqual({
            error: 'PERMISSION_DENIED',
            message:
              'You lack the permissions necessary to perform this action.',
          }),
        );
    });

    it('only remaining admin cant quit', async () => {
      const params = new URLSearchParams({
        redirect_uri: 'https://www.example.com',
        state: 'TEST_STATE',
        code_challenge: codeChallenge,
        client_id: String(1),
      });
      const codeResp = await request(app.getHttpServer())
        .post(`/auth/code?${params.toString()}`)
        .send({ email: 'TEST_USER@EMAIL.COM', password: '1oNc0iY3oml5d&%9' });
      const response = await request(app.getHttpServer())
        .post('/auth/oauth/authorization-code')
        .send({
          code: codeResp.body.code,
          redirect_uri: 'https://www.example.com',
          code_verifier: codeVerifier,
        });

      await request(app.getHttpServer())
        .delete('/platforms/1/quit')
        .set('Authorization', `Bearer ${response.body.access_token}`)
        .expect(HttpStatus.FORBIDDEN)
        .expect((res) =>
          expect(res.body).toEqual({
            error: 'NO_ADMINS_REMAINING',
            message:
              'It seems like you might be the last admin of this platform. You need to ' +
              'appoint another admin before performing this action.',
          }),
        );
    });
  });

  describe('/platforms/:platform_id/join (POST)', () => {
    beforeEach(async () => {
      await platformRepository.save(
        factories.platformEntity.build({
          redirectUris: ['https://www.example.com'],
        }),
      );
    });

    afterEach(async () => {
      await platformUserRepository.delete({});
      await platformRepository.delete({});
    });

    it('adds myself as a platform user', async () => {
      await request(app.getHttpServer())
        .post('/platforms/1/join')
        .set('Authorization', `Bearer ${userAccount.accessToken}`)
        .expect(HttpStatus.CREATED)
        .expect((res) =>
          expect(res.body).toEqual({
            id: expect.any(Number),
            profile_url: null,
            platform: {
              created_at: expect.any(String),
              id: 1,
              is_verified: true,
              name: 'TEST_PLATFORM_1',
              name_handle: 'test_platform_1#1',
              updated_at: expect.any(String),
              category: {
                id: 1,
                name: 'CATEGORY',
              },
              homepage_url: 'HOMEPAGE_URL',
            },
            roles: [UserRole.Member],
            user: {
              id: 1,
              user_handle: 'test-user#1',
              username: 'test-user',
              bio: null,
              display_name: null,
            },
          }),
        );
    });
  });
});
