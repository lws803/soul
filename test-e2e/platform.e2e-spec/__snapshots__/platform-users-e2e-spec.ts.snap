// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PlatformsController - PlatformUsers (e2e) /platforms/:platformId/join (POST) adds myself as a platform user 1`] = `
Object {
  "id": Any<Number>,
  "platform": Object {
    "category": Object {
      "id": 1,
      "name": "CATEGORY",
    },
    "created_at": Any<String>,
    "id": 1,
    "is_verified": true,
    "name": "TEST_PLATFORM_1",
    "name_handle": "test_platform_1#1",
    "updated_at": Any<String>,
  },
  "roles": Array [
    "member",
  ],
  "user": Object {
    "id": 1,
    "user_handle": "test_user#1",
    "username": "TEST_USER",
  },
}
`;

exports[`PlatformsController - PlatformUsers (e2e) /platforms/:platformId/users (GET) fetches all users within a platform 1`] = `
Object {
  "platform_users": Array [
    Object {
      "id": 1,
      "platform": Object {
        "created_at": Any<String>,
        "id": 1,
        "is_verified": true,
        "name": "TEST_PLATFORM_1",
        "name_handle": "test_platform_1#1",
        "updated_at": Any<String>,
      },
      "roles": Array [
        "admin",
        "member",
      ],
      "user": Object {
        "id": 1,
        "user_handle": "test_user#1",
        "username": "TEST_USER",
      },
    },
  ],
  "total_count": 1,
}
`;

exports[`PlatformsController - PlatformUsers (e2e) /platforms/:platformId/users/:userId (PUT) bans a user 1`] = `
Object {
  "id": 2,
  "platform": Object {
    "category": Object {
      "id": 1,
      "name": "CATEGORY",
    },
    "created_at": Any<String>,
    "id": 1,
    "is_verified": true,
    "name": "TEST_PLATFORM_1",
    "name_handle": "test_platform_1#1",
    "updated_at": Any<String>,
  },
  "roles": Array [
    "banned",
  ],
  "user": Object {
    "id": 2,
    "user_handle": "test_user_2#2",
    "username": "TEST_USER_2",
  },
}
`;

exports[`PlatformsController - PlatformUsers (e2e) /platforms/:platformId/users/:userId (PUT) sets user role 1`] = `
Object {
  "id": 2,
  "platform": Object {
    "category": Object {
      "id": 1,
      "name": "CATEGORY",
    },
    "created_at": Any<String>,
    "id": 1,
    "is_verified": true,
    "name": "TEST_PLATFORM_1",
    "name_handle": "test_platform_1#1",
    "updated_at": Any<String>,
  },
  "roles": Array [
    "admin",
    "member",
  ],
  "user": Object {
    "id": 2,
    "user_handle": "test_user_2#2",
    "username": "TEST_USER_2",
  },
}
`;
